### DISCUSSION TITLE
Cake Build system eliminate multiple cake.config files and debugging workaround with default target

### DISCUSSION THREAD

**Discussion by default-writer** [cake-build/cake#3960] [Extension requests]
Recently, i wrote an [article](https://medium.com/@hack2root/advanced-usage-of-cake-c-make-with-visual-studio-code-d7aaa3716163) about creating complex Cake Build systems recepies and created [repo](https://github.com/funcelot/cake-structured-template):

And, I stucked with nightmare because of:

1. neccessity of setting up cake.config everywhere, it has to be in each folder with scripts starting from root folder.

![image](https://user-images.githubusercontent.com/383256/188555704-0cb32e3d-2f98-483d-be47-66649703a4cb.png)

2. neccessity to add trick code with setup default targets:

![image](https://user-images.githubusercontent.com/383256/188556392-53840122-7371-484c-bba0-a23f0560f1e4.png)

 That's dubious, excessive and cause bugs and not really needed, IMHO. I believe that only one cake.config in root folder needed (the one you could get by command `git rev-parse --show-toplevel`), you do not need to define workaround code to make root-level scripts debuggable and any non-root internal tasks debuggable (at least you can add these code block on behalf of visible margins)

What do you think?

**Comment by gep13** [comment]
I am not sure I follow exactly the problem that you are having here. You haven't explained _why_ you needed to create multiple cake.config files.  What problem are you trying to solve here?

Do you have a small set of reproducible steps that can be followed to illustrate the problems that you are having?

**Reply by devlead** [reply]
`cake.config` should only be needed where the entry-level script is i.e. the `build.cake` file
https://github.com/cake-build/cake/blob/b5eea50b2a77a77ce2b9bf82f3efe3351480b162/src/Cake/Features/Building/BuildFeature.cs#L60

This means files fetched by `#load` won't need a config file along them as the config already has been loaded by that point, configuration is loaded even before the script is bootstrapped/compiled.

Also `cake.config`  only needed if your project deviates from the defaults in Cake.

Cake configuration isn't only possible through `cake.config` file, but you can also control the configuration through command-line arguments and environment variables.

For more advanced scenarios environment variables might be preferred as those are inherited by all child processed so they can be set in one place i.e. CI `yaml` or a bootsrapper shell script, or even a Cake script, i.e. in our integration tests you can see how we control Cake's configuration for child scripts through environment variables
https://github.com/cake-build/cake/blob/6d4f48c86202d2522b429933d4f60e9bb7b05396/tests/integration/Cake.Core/Scripting/AddinDirective.cake#L69-L81

**Reply by default-writer** [reply]
Hi, all!
I managed how to enable vscode to be able to do debugging and run tasks inside tree structure of cake. scripts, these cake.config and preprocessor directives in each file are essential to make all work. otherwize, you will fail to debug cakse scripts in subfolders!

@gep13 @devlead @asbjornu @frozenskys

**Reply by patriksvensson** [reply]
@funcelot No, you should only need a cake.config in the root folder, unless you use something like `CakeExecuteExpression` that @devlead mentioned.

Could you provide a minimal reproducable example of your build script?

**Comment by default-writer** [comment]
Ok, now i have this structure:

![image](https://user-images.githubusercontent.com/383256/189270543-4a31f463-d6be-4da2-9d0b-8c93bbbb151d.png)

And to be able to run and rebug task, i have to add cake.config in any of the following folders (otherwize it will crate *tools* folder in each directory)

![image](https://user-images.githubusercontent.com/383256/189273174-8924e8d0-43cd-47df-b8f1-0bfbb6e6da86.png)

without (it creates tools folder in debug on running cleanup task):

![image](https://user-images.githubusercontent.com/383256/189273594-82393c16-d7d5-4167-885e-395d1d95a16f.png)

and it still does not run:

![image](https://user-images.githubusercontent.com/383256/189273882-ef1dc14d-f437-4edf-a490-42ff84487c88.png)

And, *NOW*, after adding some glue, it is runnable and debuggable:

![image](https://user-images.githubusercontent.com/383256/189274144-ce3b4b68-3291-4210-96ba-eb46661429e9.png)

And even better, conditional targets adds glue and it allows debugging root script as well:

![image](https://user-images.githubusercontent.com/383256/189275235-b10eef30-59f0-4e35-ac91-79ecd6c30279.png)

My playground is here:

[https://github.com/funcelot/cake-structured-template](https://github.com/funcelot/cake-structured-template)

P.S.: I am working on real project, where expects big number of scrips. You may leave all scripts at root or try to run non-root scripts, but without any modifications to cake build system, you will definetely will fail, so i share with you my breakthrough path

P.P.S:

Root: cake.config

![image](https://user-images.githubusercontent.com/383256/189277189-fd2a4924-cb47-43d2-bfc2-f07b92033490.png)

Level 1 cake.config

![image](https://user-images.githubusercontent.com/383256/189277244-d199cce1-dd69-4421-932d-b0f1abb2bf61.png)

Level 2 cake.config

![image](https://user-images.githubusercontent.com/383256/189277290-0700e921-8b38-41c3-9550-f9c184022b3d.png)

Root debug cake conditional target inclusion:

![image](https://user-images.githubusercontent.com/383256/189277418-44607f52-25b8-4b62-a994-c40f0ca9f067.png)

Non-root debug cake conditional target:

![image](https://user-images.githubusercontent.com/383256/189277393-1e446416-51d1-447f-9e0d-f3f080d0e442.png)

**Reply by nils-a** [reply]
@funcelot I feel the problem here is not Cake itself, but rather how [Cake for VSCode](https://github.com/cake-build/cake-vscode) is working with multiple Cake files.

When you click "Run Task" or "Debug Task" in *Cake for VSCode*, it assumes that the one file is the "point of entry" for starting Cake (and Cake then looks for a `cake.config` next to that file). However, in a structure like yours this might be the "wrong" way to go about it.
If you'd be so kind to open an issue in [Cake for VSCode](https://github.com/cake-build/cake-vscode), someone might have the cycles to spare to think about how this can be solved.
